File: c/problem_73.c
#include "problem_73.h"
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

ListNode73 *reverse_list(ListNode73 *head) {
    ListNode73 *prev=NULL, *curr=head;
    while(curr){ ListNode73 *nxt=curr->next; curr->next=prev; prev=curr; curr=nxt; }
    return prev;
}

static ListNode73 *from_arr(int *a, int n){
    ListNode73 *d=calloc(1,sizeof(ListNode73)),*c=d;
    for(int i=0;i<n;i++){c->next=calloc(1,sizeof(ListNode73));c=c->next;c->val=a[i];}
    return d->next;
}

void problem73(void) {
    printf("Problem 73:\n");
    int a[]={1,2,3,4,5};
    ListNode73 *head=from_arr(a,5), *rev=reverse_list(head);
    int exp[]={5,4,3,2,1}, i=0;
    while(rev){assert(rev->val==exp[i++]);rev=rev->next;}
    assert(reverse_list(NULL)==NULL);
    printf("All Tests Passed\n");
}

File: rust/problem_73.rs
type Link73 = Option<Box<ListNode73>>;

pub struct ListNode73 { pub val: i32, pub next: Link73 }

impl ListNode73 {
    pub fn new(v: i32) -> Self { ListNode73 { val: v, next: None } }
}

pub fn reverse_list(mut head: Link73) -> Link73 {
    let mut prev = None;
    while let Some(mut node) = head {
        head = node.next.take();
        node.next = prev;
        prev = Some(node);
    }
    prev
}

fn from_vec(v: &[i32]) -> Link73 {
    let mut head = None;
    for &x in v.iter().rev() { let mut n = Box::new(ListNode73::new(x)); n.next = head; head = Some(n); }
    head
}

fn to_vec(mut h: &Link73) -> Vec<i32> {
    let mut r = vec![];
    while let Some(n) = h { r.push(n.val); h = &n.next; }
    r
}

pub fn problem73() {
    println!("Problem 73:");
    assert_eq!(to_vec(&reverse_list(from_vec(&[1,2,3,4,5]))), vec![5,4,3,2,1]);
    assert_eq!(to_vec(&reverse_list(from_vec(&[1]))), vec![1]);
    assert!(reverse_list(None).is_none());
    println!("All Tests Passed");
}
